components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt

```diff
@@ -20,54 +37,197 @@ typealias GeckoCrashReporter = CrashReporter
  *                [File a bug](https://bugzilla.mozilla.org/enter_bug.cgi?product=Socorro) if you would like to get your
  *                app added to the whitelist.
  */
+@Suppress("LargeClass")
 class MozillaSocorroService(
     private val applicationContext: Context,
     private val appName: String
 ) : CrashReporterService {
-    override fun report(crash: Crash.UncaughtExceptionCrash) {
-        // Not implemented currently.
+    private val logger = Logger("mozac/MozillaSocorroCrashHelperService")
+    private val startTime = System.currentTimeMillis()
+    private val defaultServerUrl = "https://crash-reports.mozilla.com/submit?" +
+            "id=${BuildConfig.MOZ_APP_ID}&version=${BuildConfig.MOZILLA_VERSION}&${BuildConfig.MOZ_APP_BUILDID}"
+    val partSet = mutableSetOf<String>()
 
-        // In theory we could upload uncaught exception crashes to Socorro too. But Socorro is not the best tool for
-        // them and it is not used by the app teams.
+    override fun report(crash: Crash.UncaughtExceptionCrash) {
+        sendReport(crash.throwable, null, null)
     }
 
     override fun report(crash: Crash.NativeCodeCrash) {
-        // GeckoView comes with a crash reporter class that we can use here. For now we are assuming that we only want
-        // to upload native code crashes to Socorro if GeckoView is used. If this assumption is going to be wrong in
-        // the future then we can start inlining the functionality here.
-        sendViaGeckoViewCrashReporter(crash)
+        sendReport(null, crash.minidumpPath, crash.extrasPath)
+    }
+
+    @Suppress("ComplexMethod", "LongMethod")
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun sendReport(throwable: Throwable?, miniDumpFilePath: String?, extrasFilePath: String?) {
+        val mURL = URL(defaultServerUrl)
+        val boundary = generateBoundary()
+        partSet.clear()
+        with(mURL.openConnection() as HttpURLConnection) {
```

Comment:
The lifetime of some of the objects is hard to follow in this large block. Do we ever close this connection (`disconnect()`)? Are we correctly closing all streams in the error case and are we catching the possible exceptions (`IOException`).

Reply:
Looked into this more.  You're correct. 

> Indicates that other requests to the server are unlikely in the near future. Calling disconnect() should not imply that this HttpURLConnection instance can be reused for other requests.

And since this is a connection that we expect there's no other requests before it times out, we should disconnect.  I'll add it.  Thanks for catching this.